liana-gui/src/export.rs

```diff
@@ -0,0 +1,386 @@
+use std::{
+    collections::HashMap,
+    fs::{self, File},
+    io::Write,
+    path::PathBuf,
+    sync::{
+        mpsc::{channel, Receiver, Sender},
+        Arc, Mutex,
+    },
+    time::{self},
+};
+
+use chrono::{DateTime, Duration, Utc};
+use liana::miniscript::bitcoin::{Amount, Txid};
+use tokio::{
+    task::{JoinError, JoinHandle},
+    time::sleep,
+};
+
+use crate::{
+    app::view,
+    daemon::{
+        model::{HistoryTransaction, Labelled},
+        Daemon, DaemonBackend, DaemonError,
+    },
+    lianalite::client::backend::api::DEFAULT_LIMIT,
+};
+
+macro_rules! send_error {
+    ($sender:ident, $error:ident) => {
+        if let Err(e) = $sender.send(ExportProgress::Error(Error::$error)) {
+            tracing::error!("ExportState::start() fail to send msg: {}", e);
+        }
+    };
+    ($sender:ident, $error:expr) => {
+        if let Err(e) = $sender.send(ExportProgress::Error($error)) {
+            tracing::error!("ExportState::start() fail to send msg: {}", e);
+        }
+    };
+}
+
+macro_rules! send_progress {
+    ($sender:ident, $progress:ident) => {
+        if let Err(e) = $sender.send(ExportProgress::$progress) {
+            tracing::error!("ExportState::start() fail to send msg: {}", e);
+        }
+    };
+    ($sender:ident, $progress:ident($val:expr)) => {
+        if let Err(e) = $sender.send(ExportProgress::$progress($val)) {
+            tracing::error!("ExportState::start() fail to send msg: {}", e);
+        }
+    };
+}
+
+#[derive(Debug, Clone)]
+pub enum ExportMessage {
+    Open,
+    ExportProgress(ExportProgress),
+    TimedOut,
+    UserStop,
+    Path(Option<PathBuf>),
+    Close,
+}
+
+impl From<ExportMessage> for view::Message {
+    fn from(value: ExportMessage) -> Self {
+        Self::Export(value)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum ExportState {
+    Init,
+    ChoosePath,
+    Path(PathBuf),
+    Started,
+    Progress(f32),
+    TimedOut,
+    Aborted,
+    Ended,
+    Closed,
+}
+
+#[derive(Debug, Clone)]
+pub enum Error {
+    Io(String),
+    HandleLost,
+    UnexpectedEnd,
+    JoinError(String),
+    ChannelLost,
+    NoParentDir,
+    Daemon(String),
+    TxTimeMissing,
```

Comment:
These final two, especially the latter, might not be relevant to all exports.

Reply:
i think it's fine for now, this + adding generics should be added when we implement a second type export imho

Reply:
Sounds good.